pkg/cloud/cloud.go

```diff
@@ -1073,3 +1087,22 @@ func getVolumeAttachmentsList(volume *ec2.Volume) []string {
 
 	return volumeAttachmentList
 }
+
+// Calculate actual IOPS for a volume and cap it at supported AWS limits.
+// Using requstedIOPSPerGB allows users to create a "fast" storage class
+// (requstedIOPSPerGB = 50 for io1), which can provide the maximum iops
+// that AWS supports for any requestedCapacityGiB.
+func capIOPS(requestedCapacityGiB int64, requstedIOPSPerGB, minTotalIOPS, maxTotalIOPS, maxIOPSPerGB int64) int64 {
+	iops := requestedCapacityGiB * requstedIOPSPerGB
+
+	if iops < minTotalIOPS {
```

Comment:
I'm not a big fan of this change. This might end up in a surprise bill for the customer, so if IOPS is smaller than minimum we should just return and explicit error instead of increasing it.

Reply:
consider the alternative as described in PR description is forcing the user to inspect the storageclass then edit their PVC to be larger which also incurs a larger bill. Using example in PR description, two scenarios:

1) We accept this PR and always increase iops to the floor: 5GiB at 100iops is 5*.125+100*.0625 = $7.125/month.
2) We force user to set the iops to the floor: 10GiB at 100 iops is  10*.125+100*.0625 = $7.75/month (+ the cost of their time to visit aws docs and pull out their calculator).

This is just one scenario, I don't really want to put the effort into drawing graphs. I think as long as we document this behavior it will be fine. Admins will appreciate they do not have to estimate their users' behavior (if they are on a tight budget they can just set iopspergb to 1 and call it a day) and users do not have to pull out their calculators.

So atm I am inclined to accept the PR.

Reply:
Here's my reasoning. If a user expects their volume to cost, say, $5 with the values they're setting, it becomes a problem at scale. The difference on paper is only $2, but what if you have 1000 of such volumes across 10 regions? Suddenly your bill is 2 x 10 x 1000 more than you expect. Am I missing something?

We can include the limits in the error message (or even a direct link to the docs) to make it a little bit easier on the user. I understand it's still not a great user experience, but I think it's preferable to having a surprise bill.

Reply:
To be safe, my idea is we don't make decision for customers, customer can adjust the iopsPerGB or volume size per their requirement as long as we deliver the message clearly (Like Ayberk mentioned, attach the doc link in the error message, etc).

Reply:
The user is mistaken in their $5 calculation if it's predicated on their volume having an impossible GB+IOPS configuration. The choice is not between $7 and $5, it's between $7 (increase both size and iops), $6 (increase only iops), and not running the Pod at all. So really we are debating whether we should let the user know about their mistake or fix it on their behalf. I prefer the latter as long as we document it, let me spell it out.

(Keep in mind on a multi-tenant cluster the person creating PVCs doesn't necessarily have permissions to create StorageClasses. Also the above numbers are imaginary but you get my point)

Say the driver provisioning for StorageClass A with iopspergb x is presented with an impossible GB+IOPS configuration. To remedy the situation, something/somebody must:

1) Increase SIZE and IOPS of volume such that its IOPS is greater than the minimum.
2) Increase IOPS of volume such that its IOPS is greater than the minimum.
3) Decide that fixing the volume to be a valid configuration would be too expensive, so do nothing, and use EFS instead or just not run their workload.

Note option 1) is ALWAYS more expensive than option 2).

If the driver throws an error and refuses to create the volume (like today), option 2 is unavailable. Because for StorageClass A there is no way for them to control iops except by increasing size of volume. So if they want to run their stateful workload they are forced to go with the more expensive option 1

If the driver fixes the volume for the user (if we accept this PR as-is), option 3 is unavailable. I don't think this option would be that popular. I think people would prefer their workload to run than not.

IF this is so contentious, we can make it a flag that defaults to current behavior.

Reply:
BTW  it seems like only bumping iops to min floor is contentious, capping iops at max is a no brainer.

Reply:
also check out this thread, i was having deja vu. https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/575 in-tree behaves this way, and I was torn on whether we should behave the same

Reply:
Well, there is CSI migration on the way and the CSI driver should behave the same as in-tree for migrated volumes. #575 mentions is going to be handled in the translation layer - but how? I don't see any code there and I don't think it's possible.

Reply:
I am open to remove increasing the IOPS, however, but it must come with a very clear error message. The current one is very confusing. With 4 GiB volume and 1 IOPS per GB, I get:

```
could not create volume in EC2: InvalidParameterValue: Volume iops of 4 is too low; minimum is 100.
```

But there is no `iops` setting in the storage class. Only `iopsPerGB`.

Reply:
talked to @msau42 on slack

What about this:
* Let's cap max IOPS to get within the supported range. I don't see any controversy here.
* For CSI storage classes, **do not** increase iops to the supported range and return error (= as it is today, only with a nicer error message).
* However, *increase the iops for storage classes migrated from in-tree*, to preserve existing behavior.

How to distinguish these two in `CreateVolume`?
  * CSI translation lib can add `alpha.csi.storage.k8s.io/migrated` parameter to CSI StorageClass when it's migrated from in-tree. Note: this translation is in-memory, in the external-provisioner. It's not stored in API server.
  * The driver sees `alpha.csi.storage.k8s.io/migrated` parameter in CreateVolume.
  * When users create a CSI StorageClass with `alpha.csi.storage.k8s.io/migrated` parameter, it's their problem. We will not document the parameter and we reserve the right to remove it in the future. That's why the `alpha` prefix, even though it will be there probably forever :-(

Reply:
> When users create a CSI StorageClass with alpha.csi.storage.k8s.io/migrated parameter, it's their problem. We will not document the parameter and we reserve the right to remove it in the future. That's why the alpha prefix, even though it will be there probably forever :-(

This also has a nice side-effect of enabling customers to bypass the error by adding this. Feels dirty, but I like it :P

I am all for this approach. It's the best of both worlds: we avoid the surprise bill issue and we don't change the behavior for transitioning volumes.

Reply:
why not make it an explicit parameter then like `ignoreIopsTooLow` (terrible name but ya)? If you introduce this hack then someone will rely on it and complain, doesn't amtter if it says alpha. migration would silently add the parameter, but for new StorageClasses admin will have the choice whether to err on side of ensuring all PVCs get a PV even if overprovisioned v.s. ensuring all PVCs either get an exactly-provisoined PV or no PV at all.

Reply:
Added parameter `allowAutoIOPSPerGBIncrease` (naming is hard!) + unit tests + documentation, PTAL.